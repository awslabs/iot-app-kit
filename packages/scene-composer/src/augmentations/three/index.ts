export * from './Anchor';
export * from './Viewpoint';
