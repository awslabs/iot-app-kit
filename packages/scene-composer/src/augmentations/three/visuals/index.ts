export * from './IVisual';
export * from './WidgetVisual';
export * from './WidgetSprite';
