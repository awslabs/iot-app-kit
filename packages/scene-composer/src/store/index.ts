export * from './StoreOperations';
export * from './Store';
