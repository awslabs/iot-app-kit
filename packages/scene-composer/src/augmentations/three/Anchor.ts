import * as THREE from 'three';

import DebugLogger from '../../logger/DebugLogger';
import { DefaultAnchorStatus, SelectedAnchor } from '../../interfaces';

import { IVisual } from './visuals';

/**
 * The generic anchor object.
 */
export class Anchor extends THREE.Object3D {
  private log = new DebugLogger('Anchor');

  protected visualMap = new Map<string, IVisual>();
  protected _visualState: DefaultAnchorStatus | string = DefaultAnchorStatus.Info;
  protected _isSelected = false;

  public add(visualWithStateName: IVisual): this {
    const stateName = visualWithStateName.name;
    if (!stateName) {
      this.log?.error('A name must be provided for the visual. This is used to determine its appearance.');
      return this;
    }

    if (!visualWithStateName.visual) {
      this.log?.error('A visual must be provided.');
      return this;
    }

    // NOTE: Add the states directly, because event bubbling to the anchor level does not seem to occur on click
    //  and is instead caught by the WidgetVisual
    super.add(visualWithStateName.visual);

    this.visualMap.set(stateName, visualWithStateName);

    // Update the anchor visual with the current visual state
    this.updateVisualState();

    return this;
  }

  public set visualState(visualState: DefaultAnchorStatus | string) {
    this._visualState = visualState;
    this.updateVisualState();
  }

  public get visualState(): DefaultAnchorStatus | string {
    return this._visualState;
  }

  /**
   * Gets the anchor's selected state
   */
  public get isSelected(): boolean {
    return this._isSelected;
  }

  /**
   * Sets the anchor's selected state keying the visibility of the specially named Selected Visual
   * @param value true if the anchor is selected
   */
  public set isSelected(value: boolean) {
    this._isSelected = value;
    this.visualMap.get(SelectedAnchor)?.setVisible(value);
  }

  private updateVisualState() {
    this.visualMap.forEach((visual, name) => {
      visual.setVisible(name === this.visualState);
    });
  }
}
