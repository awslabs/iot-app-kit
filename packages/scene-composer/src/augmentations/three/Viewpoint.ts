import * as THREE from 'three';

import DebugLogger from '../../logger/DebugLogger';

import { WidgetVisual } from './visuals';

export enum ViewpointState {
  Deselected = 'Deselected',
  Selected = 'Selected',
}

/**
 * The generic Viewpoint object.
 */
export class Viewpoint extends THREE.Object3D {
  private log = new DebugLogger('Viewpoint');

  protected visualMap = new Map<string, WidgetVisual>();
  protected _visualState: string = ViewpointState.Deselected;

  constructor() {
    super();
    this.type = 'Viewpoint';
    this.rotateX(THREE.MathUtils.degToRad(90));
  }

  // @ts-ignore
  public add(visualWithStateName: WidgetVisual): this {
    const stateName = visualWithStateName.name;
    if (!stateName) {
      this.log?.error('A name must be provided for the visual. This is used to determine its appearance.');
      return this;
    }

    if (!visualWithStateName.visual) {
      this.log?.error('A visual must be provided.');
      return this;
    }

    // NOTE: Add the states directly, because event bubbling to the anchor level does not seem to occur on click
    //  and is instead caught by the WidgetVisual
    super.add(visualWithStateName.visual);

    this.visualMap.set(stateName, visualWithStateName);

    // Update the visual with the current visual state
    this.updateVisualState();

    return this;
  }

  public set isSelected(value: boolean) {
    this._visualState = value ? ViewpointState.Selected : ViewpointState.Deselected;
    this.updateVisualState();
  }

  public get isSelected(): boolean {
    return this._visualState === ViewpointState.Selected;
  }

  private updateVisualState() {
    this.visualMap.forEach((visual, name) => {
      visual.setVisible(name === this._visualState);
    });
  }
}

export class ViewCursor extends Viewpoint {
  constructor() {
    super();
    this.type = 'ViewCursor';
  }
}
