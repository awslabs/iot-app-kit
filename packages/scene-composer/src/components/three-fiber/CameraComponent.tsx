import * as THREE from 'three';
import React, { useContext } from 'react';
import { useHelper } from '@react-three/drei/core/useHelper';
import { OrthographicCamera } from '@react-three/drei/core/OrthographicCamera';
import { PerspectiveCamera } from '@react-three/drei/core/PerspectiveCamera';
import { Camera, useFrame, useThree } from '@react-three/fiber';

import useLifcycleLogging from '../../logger/react-logger/hooks/useLifecycleLogging';
import { ISceneNodeInternal, useEditorState, ICameraComponentInternal, useSceneDocument, useStore } from '../../store';
import { sceneComposerIdContext } from '../../sceneComposerIdContext';
import { getComponentGroupName } from '../../utils/objectThreeUtils';

interface ICameraComponentProps {
  node: ISceneNodeInternal;
  component: ICameraComponentInternal;
}

export const CameraComponent: React.FC<ICameraComponentProps> = ({ node, component }: ICameraComponentProps) => {
  const sceneComposerId = useContext(sceneComposerIdContext);
  useLifcycleLogging('CameraComponent');
  const camera = React.useRef<Camera>();

  const cameraHelperRef = useHelper(camera, THREE.CameraHelper);
  const size = useThree((state) => state.size);
  const { fov, far, near } = component;
  const { document } = useSceneDocument(sceneComposerId);

  const { isEditing } = useEditorState(sceneComposerId);

  const isDefaultViewingCamera = document.defaultCameraRef?.ref === component.ref;

  // Show camera helper if in editing mode
  // TODO: it should be hidden when loading stuff
  useFrame(() => {
    if (cameraHelperRef.current) {
      cameraHelperRef.current.visible =
        useStore(sceneComposerId).getState().isEditing() && !useStore(sceneComposerId).getState().isLoadingModel;
    }
  });

  // TODO: The camera component needs to be re-designed once we add the support for viewing camera management.

  let cameraNode: JSX.Element;
  if (component.cameraType === 'Orthographic') {
    cameraNode = (
      <OrthographicCamera
        makeDefault={!isEditing() && isDefaultViewingCamera}
        // TODO: Update view frustum params
        position={new THREE.Vector3(0, 0, 0)}
        top={3}
        bottom={-3}
        left={-3}
        right={3}
        far={1000}
        zoom={1}
        near={0}
        ref={camera}
      >
        <meshBasicMaterial attach='material' />
      </OrthographicCamera>
    );
  } else {
    cameraNode = (
      <PerspectiveCamera
        makeDefault={!isEditing() && isDefaultViewingCamera}
        position={new THREE.Vector3(0, 0, 0)}
        fov={fov}
        far={far}
        near={near}
        aspect={size.width / size.height}
        ref={camera}
      >
        <meshBasicMaterial attach='material' />
      </PerspectiveCamera>
    );
  }

  return <group name={getComponentGroupName(node.ref, 'CAMERA')}>{cameraNode}</group>;
};
