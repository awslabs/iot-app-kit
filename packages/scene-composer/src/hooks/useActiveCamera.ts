import { useCallback } from 'react';
import * as THREE from 'three';

import { useSceneComposerId } from '../common/sceneComposerIdContext';
import { useEditorState } from '../store';
import { CameraControlMode, CameraSettings } from '../interfaces';
import { DEFAULT_CAMERA_OPTIONS, DEFAULT_CAMERA_POSITION, DEFAULT_CAMERA_TARGET } from '../common/constants';
import { CameraType, Transform } from '../models/SceneModels';

const useActiveCamera = () => {
  const sceneComposerId = useSceneComposerId();
  const {
    activeCameraSettings,
    setActiveCameraSettings,
    setCameraTarget,
    isViewing,
    mainCameraObject,
    activeCameraName,
    setActiveCameraName,
  } = useEditorState(sceneComposerId);

  const defaultCameraSettings: CameraSettings = {
    cameraType: CameraType.Perspective,
    fov: DEFAULT_CAMERA_OPTIONS.fov,
    far: DEFAULT_CAMERA_OPTIONS.far,
    near: DEFAULT_CAMERA_OPTIONS.near,
    zoom: 1,
    transform: {
      position: DEFAULT_CAMERA_POSITION,
    },
  };

  const calculateTargetOrGetDefault = (transform: Transform) => {
    if (transform.rotation) {
      const obj3D = new THREE.Object3D();
      obj3D.position.set(...transform.position);
      obj3D.rotation.set(...transform.rotation);
      obj3D.updateMatrix();
      obj3D.updateMatrixWorld(true);

      // Place an object to look at in front of the parent and get it's world position
      const targetObj3D = new THREE.Object3D();
      obj3D.add(targetObj3D);
      targetObj3D.position.set(0, 0, -10);
      targetObj3D.updateMatrix();
      targetObj3D.updateMatrixWorld(true);
      const target = targetObj3D.getWorldPosition(new THREE.Vector3());
      return target.toArray();
    }

    return DEFAULT_CAMERA_TARGET;
  };

  const setActiveCamera = useCallback(
    (cameraSettings?: CameraSettings, mode: CameraControlMode = 'transition', forceSetTarget = false) => {
      const settings = cameraSettings ?? defaultCameraSettings;
      setActiveCameraSettings(settings);

      const cameraTarget = {
        position: settings.transform.position,
        target: calculateTargetOrGetDefault(settings.transform),
      };

      if (forceSetTarget || isViewing()) {
        setCameraTarget(cameraTarget, mode);
      }
    },
    [setActiveCameraSettings, setCameraTarget],
  );

  return {
    activeCameraSettings,
    setActiveCameraSettings: setActiveCamera,
    mainCameraObject,
    activeCameraName,
    setActiveCameraName,
  };
};

export default useActiveCamera;
