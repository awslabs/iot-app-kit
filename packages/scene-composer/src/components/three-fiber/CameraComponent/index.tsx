import * as THREE from 'three';
import React, { useEffect } from 'react';
import { useHelper } from '@react-three/drei/core/useHelper';
import { OrthographicCamera, PerspectiveCamera } from '@react-three/drei';
import { Camera, useThree } from '@react-three/fiber';

import useLifecycleLogging from '../../../logger/react-logger/hooks/useLifecycleLogging';
import { ISceneNodeInternal, useEditorState, ICameraComponentInternal } from '../../../store';
import { useSceneComposerId } from '../../../common/sceneComposerIdContext';
import { getComponentGroupName } from '../../../utils/objectThreeUtils';
import useSelectedNode from '../../../hooks/useSelectedNode';
import useActiveCamera from '../../../hooks/useActiveCamera';
import { getCameraSettings } from '../../../utils/cameraUtils';

interface ICameraComponentProps {
  node: ISceneNodeInternal;
  component: ICameraComponentInternal;
}

const CameraComponent: React.FC<ICameraComponentProps> = ({ node, component }: ICameraComponentProps) => {
  const sceneComposerId = useSceneComposerId();
  useLifecycleLogging('CameraComponent');
  const camera = React.useRef<Camera>();

  const cameraHelperRef = useHelper(camera, THREE.CameraHelper);
  const size = useThree((state) => state.size);
  const { fov, far, near, zoom } = component;

  const { isEditing, isLoadingModel, getObject3DBySceneNodeRef } = useEditorState(sceneComposerId);

  const { selectedSceneNodeRef } = useSelectedNode();
  const { activeCameraName, setActiveCameraSettings } = useActiveCamera();

  useEffect(() => {
    if (activeCameraName === node.name) {
      const object3D = getObject3DBySceneNodeRef(node.ref);

      setActiveCameraSettings(
        getCameraSettings(object3D, {
          cameraType: 'Perspective',
          fov,
          near,
          far,
          zoom,
        }),
      );
    }
  }, [activeCameraName]);

  useEffect(() => {
    if (cameraHelperRef.current) {
      cameraHelperRef.current.visible = isEditing() && !isLoadingModel;
    }
  }, [isEditing, isLoadingModel, cameraHelperRef]);

  let cameraNode: JSX.Element;
  if (component.cameraType === 'Orthographic') {
    cameraNode = (
      <OrthographicCamera
        // TODO: Make Editable once we expose this camera type
        top={3}
        bottom={-3}
        left={-3}
        right={3}
        far={far}
        zoom={1}
        near={near}
        ref={camera}
      >
        <meshBasicMaterial attach='material' />
      </OrthographicCamera>
    );
  } else {
    cameraNode = (
      <PerspectiveCamera fov={fov} far={far} near={near} aspect={size.width / size.height} ref={camera}>
        <meshBasicMaterial attach='material' />
      </PerspectiveCamera>
    );
  }

  return (
    <group name={getComponentGroupName(node.ref, 'CAMERA')}>
      {isEditing() && selectedSceneNodeRef === node.ref && cameraNode}
    </group>
  );
};

export default CameraComponent;
