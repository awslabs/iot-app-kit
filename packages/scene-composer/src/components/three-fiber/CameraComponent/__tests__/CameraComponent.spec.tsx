import React from 'react';
import * as THREE from 'three';
import { render } from '@testing-library/react';
import { useThree } from '@react-three/fiber';

import { ICameraComponentInternal, ISceneNodeInternal, useEditorState, accessStore } from '../../../../store';
import useActiveCamera from '../../../../hooks/useActiveCamera';
import CameraComponent from '..';

import Mock = jest.Mock;

jest.mock('../../../../store', () => {
  const originalModule = jest.requireActual('../../../../store');
  return {
    ...originalModule,
    accessStore: jest.fn(),
    useSceneDocument: jest.fn(() => ({ document: { defaultCameraRef: { ref: 'testCamera' } } })),
    useEditorState: jest.fn(),
  };
});

jest.mock('@react-three/fiber', () => {
  const originalModule = jest.requireActual('@react-three/fiber');
  return {
    ...originalModule,
    useThree: jest.fn(),
    useFrame: jest.fn().mockImplementation((callback) => callback()),
  };
});

jest.mock('../../../../hooks/useSelectedNode', () => {
  return jest.fn().mockReturnValue({
    selectedSceneNodeRef: 'testRef',
  });
});

jest.mock('../../../../hooks/useActiveCamera', () => {
  return jest.fn().mockReturnValue({ activeCameraName: 'test-camera', setActiveCameraName: jest.fn() });
});

jest.mock('../../../../hooks/useEditorHelper', () => {
  return { useEditorHelper: jest.fn() };
});

describe('CameraComponent', () => {
  const node = {
    ref: 'testRef',
    name: 'testCamera',
    transform: {
      position: [5, 5, 5],
    },
  } as ISceneNodeInternal;
  const isEditing = jest.fn();
  const isViewing = jest.fn();
  const object3D = new THREE.Object3D();
  object3D.position.set(5, 5, 5);
  object3D.rotation.set(0, 0, 0);
  object3D.scale.set(1, 1, 1);

  beforeEach(() => {
    const mockState = { isEditing: true, isLoadingModel: false };

    const accessStoreMock = accessStore as Mock;
    accessStoreMock.mockReturnValue({ getState: jest.fn(() => mockState) });

    const useThreeMock = useThree as Mock;
    useThreeMock.mockReturnValue({ width: 1920, height: 1080 });

    const useEditorStateMock = useEditorState as Mock;
    useEditorStateMock.mockReturnValue({
      isEditing,
      isViewing,
      getObject3DBySceneNodeRef: jest.fn().mockReturnValue(object3D),
    });

    jest.clearAllMocks();
  });

  it('should render correctly for perspective camera', () => {
    isEditing.mockReturnValue(true);
    (useActiveCamera as jest.Mock).mockReturnValue({
      activeCameraSettings: {
        cameraType: 'Perspective',
        fov: 100,
        far: 50,
        near: 0.2,
        transform: {
          position: [5, 5, 5],
        },
      },
      setActiveCameraSettings: jest.fn(),
    });
    const component = {
      ref: 'testCamera',
      fov: 60,
      near: 0,
      far: 100,
      zoom: 1,
      cameraType: 'Perspective',
    } as ICameraComponentInternal;

    const { container } = render(<CameraComponent node={node} component={component} />);
    expect(container).toMatchSnapshot();
    expect(useActiveCamera().setActiveCameraSettings).not.toBeCalled();
  });

  it('should render as default correctly for perspective camera in viewing mode', () => {
    isEditing.mockReturnValue(false);
    const component = {
      ref: 'testCamera',
      fov: 60,
      near: 0,
      far: 100,
      zoom: 1,
      cameraType: 'Perspective',
    } as ICameraComponentInternal;

    const { container } = render(<CameraComponent node={node} component={component} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly for orthographic camera', () => {
    isEditing.mockReturnValue(true);
    const component = {
      ref: 'testCamera',
      fov: 60,
      near: 0,
      far: 100,
      zoom: 1,
      cameraType: 'Orthographic',
    } as ICameraComponentInternal;

    const { container } = render(<CameraComponent node={node} component={component} />);
    expect(container).toMatchSnapshot();
  });

  it('should call setActiveCameraSettings if activeCameraName matches the name of the node', () => {
    isEditing.mockReturnValue(true);
    (useActiveCamera as jest.Mock).mockReturnValue({
      activeCameraSettings: {
        cameraType: 'Perspective',
        fov: 100,
        far: 50,
        near: 0.2,
        transform: {
          position: [5, 5, 5],
        },
      },
      setActiveCameraSettings: jest.fn(),
      activeCameraName: 'testCamera',
    });
    const component = {
      ref: 'testCamera',
      fov: 60,
      near: 0,
      far: 100,
      zoom: 1,
      cameraType: 'Perspective',
    } as ICameraComponentInternal;

    render(<CameraComponent node={node} component={component} />);

    const argValue = (useActiveCamera().setActiveCameraSettings as jest.Mock).mock.calls[0][0];

    expect(argValue).toMatchInlineSnapshot(`
      {
        "cameraType": "Perspective",
        "far": 100,
        "fov": 60,
        "near": 0,
        "transform": {
          "position": [
            5,
            5,
            5,
          ],
          "rotation": [
            -0,
            0,
            -0,
          ],
          "scale": [
            1,
            1,
            1,
          ],
        },
        "zoom": 1,
      }
    `);
  });
});
