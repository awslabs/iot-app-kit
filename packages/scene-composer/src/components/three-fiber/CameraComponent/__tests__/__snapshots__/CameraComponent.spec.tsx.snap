// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CameraComponent should render as default correctly for perspective camera in viewing mode 1`] = `
<div>
  <group
    name="CAMERA_COMPONENT_testRef"
  />
</div>
`;

exports[`CameraComponent should render correctly for orthographic camera 1`] = `
<div>
  <group
    name="CAMERA_COMPONENT_testRef"
  >
    <div
      bottom="-3"
      data-mocked="OrthographicCamera"
      far="100"
      left="-3"
      near="0"
      right="3"
      top="3"
      zoom="1"
    >
      <meshbasicmaterial
        attach="material"
      />
    </div>
  </group>
</div>
`;

exports[`CameraComponent should render correctly for perspective camera 1`] = `
<div>
  <group
    name="CAMERA_COMPONENT_testRef"
  >
    <div
      aspect="1.7777777777777777"
      data-mocked="PerspectiveCamera"
      far="100"
      fov="60"
      near="0"
    >
      <meshbasicmaterial
        attach="material"
      />
    </div>
  </group>
</div>
`;
